### PR TITLE
Locking particular alert in etcd while reading and updating

### PR DESCRIPTION
**Pull request by GowthamShanmugam**
tendrl-bugid: Tendrl/notifier#139

Signed-off-by: GowthamShanmugam <gshanmug@redhat.com>

### CODE REVIEW

**Comment by GowthamShanmugam** [comment]
@shtripat @r0h4n please review

**Comment by GowthamShanmugam** [comment]
tested

File: tendrl/notifier/notification/__init__.py [hunk 1]
```diff
@@ -118,6 +118,12 @@ def run(self):
                     NS.config.data["notification_check_interval"]
                 )
                 time.sleep(interval)
+                lock = etcd.Lock(NS._int.client, 'alerting')
+                lock.acquire(blocking=True,
+                             lock_ttl=60)
+                if lock.is_acquired:
+                    # renew a lock
+                    lock.acquire(lock_ttl=60)
```
Comment (shtripat) [inline tendrl/notifier/notification/__init__.py]: Why need renewal of lock?

Comment (GowthamShanmugam) [reply tendrl/notifier/notification/__init__.py]: Let's take example node A and Node B are the two machines which will try to push the alerts. If node A acquires the lock and proceeding further, So at same time node B also trying but it unable to acquire lock so it has to wait for some time. When node A finishes then node B acquired lock but it already lost some time on waiting. So renewing the lock will help to give another 60 sec to finish the node B job.

Comment (shtripat) [reply tendrl/notifier/notification/__init__.py]: Ack. Changes look fine

File: tendrl/notifier/notification/__init__.py [hunk 2]
```diff
@@ -122,9 +123,21 @@ def run(self):
                 for alert in alerts:
                     alert.tags = json.loads(alert.tags)
                     if alert.delivered == "False":
+                        lock = etcd.Lock(
+                            NS._int.client,
```
Comment (r0h4n) [inline tendrl/notifier/notification/__init__.py]: use NS._int.wclient  for locking. How is locking the etcd read client going to help?

Comment (GowthamShanmugam) [reply tendrl/notifier/notification/__init__.py]: i will change it

Comment (GowthamShanmugam) [reply tendrl/notifier/notification/__init__.py]: correct, my mistake i will change this

**Comment by GowthamShanmugam** [comment]
@r0h4n changes done

File: tendrl/notifier/notification/__init__.py [hunk 3]
```diff
@@ -122,9 +123,21 @@ def run(self):
...
                     if alert.delivered == "False":
+                        lock = etcd.Lock(
+                            NS._int.wclient,
+                            'alerting/alerts/%s' % alert.alert_id
+                        )
+                        lock.acquire(
+                            blocking=True,
+                            lock_ttl=60
+                        )
+                    if lock.is_acquired:
```
Comment (r0h4n) [inline tendrl/notifier/notification/__init__.py]: this if condition is outside the scope in which the "lock" is created and acquired,

youll need to move it inside the scope, current code will fail if alert.delivered != "False"

Comment (GowthamShanmugam) [reply tendrl/notifier/notification/__init__.py]: yes, i need to move why it is not failed and works fine means, an exception is caught by exception handler and for it worked for undelivered alerts fine. i will change this

**Comment by GowthamShanmugam** [comment]
@r0h4n changes done
